language/metrics-output-plugin/src/__tests__/metrics-output-plugin.test.ts

```diff
@@ -0,0 +1,350 @@
+import { test, expect, describe, beforeEach, afterEach, vi } from "vitest";
+import * as fs from "fs";
+import * as path from "path";
+import { TextDocument } from "vscode-languageserver-textdocument";
+import {
+  ReferenceAssetsWebPluginManifest,
+  Types,
+} from "@player-tools/static-xlrs";
+import { PlayerLanguageService } from "@player-tools/json-language-service";
+
+import { MetricsOutput, extractFromDiagnostics, extractBySource } from "..";
+import { ComplexityCheck } from "@player-tools/complexity-check-plugin";
+
+describe("WriteMetricsPlugin", () => {
+  let service: PlayerLanguageService;
+  const TEST_DIR = path.resolve("target");
+  const TEST_FILE = "output.json";
+  const TEST_FILE_PATH = path.join(TEST_DIR, TEST_FILE);
+
+  const mockAssetCount = `{
+      "action": 3,
+      "text": 3,
+      "collection": 2,
+      "image": 2,
+      "input": 2,
+      "info": 1
+  }`;
+
+  beforeEach(async () => {
+    // Create test directory
+    if (!fs.existsSync(TEST_DIR)) {
+      fs.mkdirSync(TEST_DIR, { recursive: true });
+    }
+
+    // Set fixed date for consistent test results
+    vi.spyOn(Date.prototype, "toISOString").mockImplementation(
+      () => "2023-01-01T00:00:00.000Z",
+    );
+
+    service = new PlayerLanguageService();
+
+    // Add complexity check plugin first
+    service.addLSPPlugin(
+      new ComplexityCheck({
+        maxAcceptableComplexity: 60,
+      }),
+    );
+
+    // Add a custom diagnostic plugin to report asset count
+    service.addLSPPlugin({
+      name: "AssetCountDiagnostic",
+      apply(languageService: PlayerLanguageService) {
+        // Hook into the validate event
+        languageService.hooks.validate.tap(
+          "AssetCountDiagnostic",
+          async (documentContext, validationContext) => {
+            validationContext.addDiagnostic({
+              message: `${mockAssetCount}`,
+              range: {
+                start: { line: 0, character: 0 },
+                end: { line: 0, character: 1 },
+              },
+              severity: 1, // Information
+              source: "asset-count",
+            });
+          },
+        );
+      },
+    });
+
+    service.addLSPPlugin(
+      new MetricsOutput({
+        outputDir: TEST_DIR,
+        fileName: TEST_FILE.replace(".json", ""),
+        rootProperties: {
+          testRun: true,
+        },
+        stats: {
+          complexity: extractFromDiagnostics(
+            /Content complexity is (\d+)/,
+            (value: string) => parseInt(value, 10),
+          ),
+          assets: (diagnostics) => extractBySource("asset-count", diagnostics),
```

Comment:
String based lookups are probably fine but IMO we should:
- Properly document the potential issues around namespace collision
- Offer a way to use a Symbol as the look up.